subprojects/plugins/src/integTest/groovy/org/gradle/groovy/compile/GroovyCompileToolchainIntegrationTest.groovy

```diff
@@ -0,0 +1,205 @@
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.testing.fixture.GroovyCoverage
+import org.gradle.util.internal.TextUtil
+
+import static org.gradle.util.internal.GroovyDependencyUtil.groovyModuleDependency
+
+@TargetCoverage({ GroovyCoverage.SINCE_3_0 })
+class GroovyCompileToolchainIntegrationTest extends MultiVersionIntegrationSpec implements JavaToolchainFixture {
+
+    def setup() {
+        file("src/main/groovy/Foo.java") << "public class Foo {}"
+        file("src/main/groovy/Bar.groovy") << "public class Bar { def bar() {} }"
+
+        buildFile << """
+            apply plugin: "groovy"
+            ${mavenCentralRepository()}
+
+            dependencies {
+                implementation "${groovyModuleDependency("groovy", version)}"
+            }
+        """
+    }
+
+    def "forkOptions #option is ignored"() {
+        def currentJdk = Jvm.current()
+        def otherJdk = AvailableJavaHomes.getDifferentVersion()
+
+        if (option == "executable") {
+            buildFile << """
+                compileGroovy {
+                    options.fork = true
+                    options.forkOptions.executable = "${TextUtil.normaliseFileSeparators(otherJdk.javaExecutable.absolutePath)}"
+                }
+            """
+        } else {
+            buildFile << """
+                compileGroovy {
+                    options.fork = true
+                    options.forkOptions.javaHome = file("${TextUtil.normaliseFileSeparators(otherJdk.javaHome.absolutePath)}")
+                }
+            """
+        }
+
+        def groovyTarget = GroovyCoverage.getEffectiveTarget(versionNumber, currentJdk.javaVersion)
+
+        when:
+        withInstallations(otherJdk).run(":compileGroovy")
+
+        then:
+        executedAndNotSkipped(":compileGroovy")
+        currentJdk.javaVersion == JavaVersion.forClass(groovyClassFile("Foo.class").bytes)
+        groovyTarget == JavaVersion.forClass(groovyClassFile("Bar.class").bytes)
+
+        where:
+        option << ["executable", "javaHome"]
+    }
+
+    def "uses #what toolchain #when"() {
+        def currentJdk = Jvm.current()
+        def otherJdk = AvailableJavaHomes.differentVersion
+        def selectJdk = { it == "other" ? otherJdk : it == "current" ? currentJdk : null }
+
+        if (withTool != null) {
+            configureTool(selectJdk(withTool))
+        }
+        if (withJavaExtension != null) {
+            configureJavaPluginToolchainVersion(selectJdk(withJavaExtension))
+        }
+
+        def targetJdk = selectJdk(target)
+        def groovyTarget = GroovyCoverage.getEffectiveTarget(versionNumber, targetJdk.javaVersion)
+
+        when:
+        withInstallations(currentJdk, otherJdk).run(":compileGroovy", "--info")
+
+        then:
+        executedAndNotSkipped(":compileGroovy")
+        outputContains("Compiling with JDK Java compiler API")
+        targetJdk.javaVersion == JavaVersion.forClass(groovyClassFile("Foo.class").bytes)
```

Comment:
```suggestion
        JavaVersion.forClass(groovyClassFile("Foo.class").bytes) == targetJdk.javaVersion
```

💄  I wonder if this should be turned around, given that the `targetJdk.javaVersion` is the expected value. Same for the other assertions like this.

Reply:
Done